query-service-impl/src/main/java/org/hypertrace/core/query/service/ExecutionContext.java

```diff
@@ -35,14 +38,83 @@
   // while the selectedColumns
   // is a set of column names.
   private final LinkedHashSet<Expression> allSelections;
+  private final Map<String,String> timeSeriesColumnMap;
+  private final Map<String,List<Long>> timeFilterMap;
 
   public ExecutionContext(String tenantId, QueryRequest request) {
     this.tenantId = tenantId;
     this.selectedColumns = new LinkedHashSet<>();
     this.allSelections = new LinkedHashSet<>();
+    this.timeSeriesColumnMap = new HashMap<>();
+    this.timeFilterMap = new HashMap<String,List<Long>>();
+    analyzeForAvgRate(request);
     analyze(request);
   }
 
+  private void analyzeForAvgRate(QueryRequest request){
+    setTimeSeriesColumnMap(request);
+    setTimeFilterMap(request);
+  }
+
+  private void setTimeSeriesColumnMap(QueryRequest request){
+
+    if(request.getGroupByCount() > 0){
+      for(Expression expression : request.getGroupByList()){
+        if(expression.getValueCase() == ValueCase.FUNCTION && expression.getFunction().getFunctionName().equals("dateTimeConvert")) {
+
+          String column = null;
+          String period = null;
+
+          //assuming only one column and one literal (with period in seconds) is present in one dateTimeConvert groupBy
+          for(Expression childExpression : expression.getFunction().getArgumentsList()) {
+            if(childExpression.getValueCase() == ValueCase.COLUMNIDENTIFIER) {
+              column = childExpression.getColumnIdentifier().getColumnName().split("[.]")[0];
+            }
+            if(childExpression.getValueCase() == ValueCase.LITERAL){
+              if(childExpression.getLiteral().getValue().getString().split("[:]")[1].equals("SECONDS")){
+                period = childExpression.getLiteral().getValue().getString().split("[:]")[0];
+              }
+            }
+          }
+
+          if(column != null && period != null){
+            timeSeriesColumnMap.put(column,period);
```

Comment:
So what are we trying to get out of this now that we've found that col X is being grouped in buckets of Y seconds, and how can we make it easiest for callers?

At least right now, we don't care about the column name as far as I can tell, so we can probably discard it (we may need to bring it back in the future). I went and looked in the other file to see how it's being used there, and it actually looks like a bug anyway - we're using the argument of the rate function to match against this map, but those are two different columns.

My suggestion - convert period into a duration to remove unit and assumptions, then expose it via `Optional<Duration> getTimeSeriesInterval()`

Reply:
We can make this assumption now as there will be only one time filter (lg and gt)  and one dataTimeConvert function for a given query. So, execution context can have two vars - 1. `getTimeSeriesInterval()` and 2. `getTimeInterval()`